URL generation for named routes

aaronkavlie-wf opened the pull request:
Problem
=======

It would be useful to be able to generate a URL for a given endpoint with a set of parameters -- for example, in a POST endpoint, you might want to provide the user with the GET endpoint where they can retrieve the resource they just created.
 
See webapp2's [uri_for()](https://webapp-improved.appspot.com/guide/routing.html#guide-routing-building-uris) among others for an example of this.

Solution
======

Added the router to the `gorillaRequestContext` struct, and `BuildURL()` & `BuildPath()` methods to the `RequestContext` interface, so the user can get a full URL or path for an endpoint.

Code Review
===========
@tylertreat 
@alexandercampbell-wf 

FYI: 
@trentonsmith-wf

alexandercampbell-wf commented on rest/context.go:
```diff
@@ -325,6 +329,30 @@ func (ctx *gorillaRequestContext) NextURL() (string, error) {
+	if err != nil {
+		log.Println(err)
+	}
+	return fmt.Sprintf("%v", uri), nil
```
`Sprintf` is expensive because it has to parse the formatting string and then apply reflect to its arguments. It's not doing anything here anyway. Instead, call `uri.String()` directly.

alexandercampbell-wf commented on rest/context.go:
```diff
@@ -325,6 +329,30 @@ func (ctx *gorillaRequestContext) NextURL() (string, error) {
+		URL("version", ctx.Version(),
+		"resource_id", resourceID)
+	if err != nil {
+		log.Println(err)
```
Not returning the error? I don't think `uri` will be valid if `err != nil`.

alexandercampbell-wf commented on rest/context.go:
```diff
@@ -72,6 +73,8 @@ type RequestContext interface {
 	// string is returned with the error set.
 	NextURL() (string, error)
 
+	BuildURL(string, string) (string, error)
```
Please provide a docstring beginning with the name of the function like the other methods on the interface. I'd also recommend naming the parameters. From this definition I have no idea how to use the function.

aaronkavlie-wf replied:
Yeah, named parameters would help... I omitted them because nothing else on the interface has them.

alexandercampbell-wf commented on rest/context.go:
```diff
@@ -130,12 +133,13 @@ type RequestContext interface {
 // NewContext returns a RequestContext populated with parameters from the request path and
 // query string.
-func NewContext(parent context.Context, req *http.Request) RequestContext {
+func NewContext(parent context.Context, req *http.Request, router *mux.Router) RequestContext {
```
This signature change breaks a lot of existing code. I haven't attempted a fix for any repos, so it may be a trivial issue. Just something to recognize. Some devs outside of Workiva will probably see broken builds as well.

aaronkavlie-wf replied:
I thought this would be relatively safe as it isn't directly set by the user within handlers. I do see a lot of calls in gotable-server tests though.

We could always add a `NewContextWithRouter` or something and preserve the original.

tylertreat-wf replied:
Yeah, would prefer to minimize breaking API changes if possible.

alexandercampbell-wf commented on rest/context.go:
```diff
@@ -325,6 +329,30 @@ func (ctx *gorillaRequestContext) NextURL() (string, error) {
+func (ctx *gorillaRequestContext) BuildURL(routeName, resourceID string) (string, error) {
+	r, ok := ctx.Request()
+	if !ok {
+		return "", fmt.Errorf("Unable to build next url: no request")
```
This can just be `errors.New()` since you're not using any formatting parameters. However, I'd prefer this had scope and detail.

```go
fmt.Errorf(`go-rest: unable to build the next url for route %q, resource %q`,
        routeName, resourceID)
```

aaronkavlie-wf replied:
I copied that portion directly from [NextURL()](https://github.com/Workiva/go-rest/blob/master/rest/context.go#L305-308) without much thought  :-)

Will update, though the error message should be clear that the problem is the lack of a request on the context, not the arguments.

aaronkavlie-wf replied:
Oh, and I didn't know about `%q`, that's handy!

alexandercampbell-wf commented:
I can't do much beyond reviewing the syntax, as REST libraries were never my strong suit :) @tylertreat-wf will know more regarding your discussion points.

aaronkavlie-wf commented:
@alexandercampbell-wf thanks for the feedback, addressed your comments.

alexandercampbell-wf commented on rest/context.go:
```diff
@@ -325,6 +329,32 @@ func (ctx *gorillaRequestContext) NextURL() (string, error) {
+	r, ok := ctx.Request()
+	if !ok {
+		return "", fmt.Errorf("Unable to build URL for route name %q with resource ID %q: No request available.",
+			routeName, resourceID)
```
Error messages shouldn't be capitalized or end with punctuation.

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

alexandercampbell-wf commented:
+1 pending the addition of unit tests.

tylertreat-wf commented on rest/api.go:
```diff
@@ -159,6 +159,8 @@ type API interface {
 	// format is not implemented, the returned serializer will be nil and the error set.
 	responseSerializer(string) (ResponseSerializer, error)
+
+	Router() *mux.Router
```
I don't like the idea of exposing `mux` on the `API` interface. This interface is meant to be generic. `mux` is just one specific backing implementation. I would prefer to avoid exposing this altogether, but if we must, make it an interface or something which `mux` can implement.

aaronkavlie-wf replied:
It's exposed here solely to make it available [in the requestHandler methods](https://github.com/aaronkavlie-wf/go-rest/blob/build_url/rest/handler.go#L134). Is there another way to get from here to there without exposing the `mux.Router`?

tylertreat-wf replied:
Could we just add it directly to the `requestHandler` struct?

aaronkavlie-wf replied:
That works, just pushed a commit with that.

tylertreat-wf commented on rest/context.go:
```diff
@@ -72,6 +72,9 @@ type RequestContext interface {
 	NextURL() (string, error)
 
+	// BuildURL builds the URL for the given route name and resource ID.
+	BuildURL(routeName, resourceID string) (string, error)
```
Instead of passing a route name, could we just have constants like `Get`, `GetList`, `Create`, etc.? This would make sense to the user I think, then we could handle the internal route name without leaking it.

aaronkavlie-wf replied:
Yeah, I think a set of constants makes sense.

aaronkavlie-wf commented:
@alexandercampbell-wf @tylertreat-wf extensively revised & polished this; it's ready for another look.

alexandercampbell-wf commented on rest/context.go:
```diff
@@ -325,6 +351,60 @@ func (ctx *gorillaRequestContext) NextURL() (string, error) {
+	methodName string, pairs ...string) (string, error) {
+	r, ok := ctx.Request()
+	if !ok {
+		return "", fmt.Errorf("unable to build URL for resource name %q: no request available.",
```
Minor comment, but this error message shouldn't end with a period.

alexandercampbell-wf commented on rest/context.go:
```diff
@@ -325,6 +351,60 @@ func (ctx *gorillaRequestContext) NextURL() (string, error) {
+	routeName := resourceName + ":" + methodName
+	route := ctx.router.Get(routeName).Host(r.Host)
+
+	var builder func(pairs ...string) (*url.URL, error)
```
I'm not sure I like this `pairs` system. Is this consistent with another piece of go-rest, or something new?

aaronkavlie-wf replied:
That's passed straight through to [gorilla's Route.URL() method](http://www.gorillatoolkit.org/pkg/mux#Route.URL).

I'm not in love with it either, but I'm not sure if the alternatives are any better in Go land. I considered a map, which would be a decent solution if not for the `map[string]string{}` line noise necessary on usage.

Keyword args sure would be nice...

alexandercampbell-wf replied:
Theoretically you could declare a type alias such as

```go
type pairs map[string]string
```

and then use it like this:

```go
var p = pairs{"option1": "setting1", "option2": "setting2"}
```

But it's really @tylertreat-wf's call whether that makes sense for this library.

aaronkavlie-wf replied:
Users would have to include the package name, so it would really be `rest.pairs`:

```go
url, err := rest.BuildURL("resource", rest.HandleCreate, rest.pairs{"foo": "bar"})
```

still less off-putting than `map[string]string{}` at least... I'd like to hear what Tyler thinks.

tylertreat-wf replied:
Yes, I think I prefer Alex's suggestion over the list of strings. It's an annoying choice gorilla made, but we can at least prevent leaking it into our API.

alexandercampbell-wf commented on rest/context.go:
```diff
@@ -325,6 +351,60 @@ func (ctx *gorillaRequestContext) NextURL() (string, error) {
+// corresponds with the CreateResource method.
+//
+// pairs should be key/value pairs for any URL variables, passed on to
+// gorilla's Route.URL method.
```
This docstring should be specific to the implementation rather than a copy/paste from the interface.

aaronkavlie-wf replied:
Actually, I copy/pated *from* here to the interface.

It's not totally clear to me how the interface/implementation docs should differ; a lot of the existing docstrings are copy/pasted.

alexandercampbell-wf replied:
I hate, *hate*, **hate** any duplication in source code. I'd rather the comment on the implementation be deleted rather than copy/pasted from the interface.

@tylertreat-wf thoughts?

tylertreat-wf replied:
I tend to copy the docstring from the interface to the implementation because I find myself looking at the implementation more. It's pretty annoying having to find the interface to figure out what a method should do.

alexandercampbell-wf commented on rest/context.go:
```diff
@@ -325,6 +351,60 @@ func (ctx *gorillaRequestContext) NextURL() (string, error) {
+// BuildPath builds a URL path for the given resource name & method name with the
+// key/value pairs for the route variables.
+//
+// See BuildURL for more details.
```
Same comment here.

alexandercampbell-wf commented on rest/context_test.go:
```diff
@@ -97,3 +132,35 @@ func TestHeader(t *testing.T) {
+	api.RegisterResourceHandler(ComplexTestResourceHandler{})
+
+	req, _ := http.NewRequest("GET", "https://example.com/api/v1/widgets", nil)
+	gContext.Set(req, "version", "1")
```
I don't see any versions other that `v1`. Would it make sense to add coverage over other versions like `v2` in this test?

alexandercampbell-wf replied:
Coverage added in https://github.com/aaronkavlie-wf/go-rest/commit/b1bd6a15687b42b039987311c58d3938806c94be

tylertreat-wf commented on rest/api.go:
```diff
@@ -30,6 +30,14 @@ import (
+	HandleUpdate     = "update"
+	HandleDelete     = "delete"
+	HandleReadList   = "readList"
+	HandleUpdateList = "updateList"
```
I would actually typedef these.

```go
type HandleMethod string

const (
        HandleCreate HandleMethod = "create"
        // etc...
)
```

tylertreat-wf commented on rest/context.go:
```diff
@@ -72,6 +72,25 @@ type RequestContext interface {
+	//
+	// pairs should be key/value pairs for any URL variables, passed on to
+	// gorilla's Route.URL method.
+	BuildURL(resourceName string, methodName string, pairs ...string) (string, error)
```
See comment above about typedef for method name. Then this would look like:

```go
BuildURL(resourceName string, method HandleMethod, pairs ...string) (string, error)
```

It's more clear to the user what valid values of `methodName` are.

tylertreat-wf commented on rest/context.go:
```diff
@@ -72,6 +72,25 @@ type RequestContext interface {
+	// corresponds with the CreateResource method.
+	//
+	// pairs should be key/value pairs for any URL variables, passed on to
+	// gorilla's Route.URL method.
```
gorilla is an implementation detail.

tylertreat-wf commented on rest/context.go:
```diff
@@ -72,6 +72,25 @@ type RequestContext interface {
+	// key/value pairs for the route variables.
+	//
+	// See BuildURL for more details.
+	BuildPath(resourceName string, methodName string, pairs ...string) (string, error)
```
See above comment.

aaronkavlie-wf commented:
@tylertreat-wf @alexandercampbell-wf addressed code review comments.

Adopted API suggestions, but used the name `RouteVars` for the map of route variables.

Still duplicating docstrings for interface and implementation; if you have a suggestion there let me know.

I agree that we should avoid repetition; I spend most of the time looking at implementations as Tyler said. Godoc formatting also makes the implementation docs more readable, so I think that's the ideal place for the full docstring.

alexandercampbell-wf commented on rest/context.go:
```diff
@@ -325,6 +348,64 @@ func (ctx *gorillaRequestContext) NextURL() (string, error) {
 	return u.String(), nil
 }
 
+type RouteVars map[string]string
```
Please add a docstring explaining why this exists and for what it is used. It would be nice to have an example too, I've already forgotten the details of what this looks like :)

alexandercampbell-wf replied:
(if I was unclear, I meant an example RouteVars declaration in the docstring)

aaronkavlie-wf replied:
docstring added!

tylertreat-wf commented on rest/api.go:
```diff
@@ -27,9 +27,19 @@ import (
+	HandleUpdate     HandleMethod = "update"
+	HandleDelete     HandleMethod = "delete"
+	HandleReadList   HandleMethod = "readList"
+	HandleUpdateList HandleMethod = "updateList"
```
This is a nit, but you just need to put `HandleMethod` on the first one and it will be implied for the rest.

```go
	HandleCreate     HandleMethod = "create"
	HandleRead                    = "read"
	HandleUpdate                  = "update"
	HandleDelete                  = "delete"
	HandleReadList                = "readList"
	HandleUpdateList              = "updateList"
```

aaronkavlie-wf replied:
fixed

tylertreat-wf commented:
+1

alexandercampbell-wf commented:
+1

tylertreat-wf commented:
+1

tylertreat-wf commented:
@stevenosborne-wf @wesleybalvanz-wf

stevenosborne-wf commented:
+1
